### PR TITLE
Modernize flock: CI, bug fixes, comprehensive test suite, and refactoring

### LINKED ISSUE #5: convert tests to not be time sensitive
time-sensitive tests do not perform well on extremely small virtual machines. Probably they should be modified to check order of execution.

### LINKED ISSUE #15: Add -h/--help option
Would be nice to have a `-h`/`--help` option that returns with exit status 0.

### LINKED ISSUE #17: travis tests for linux
At the very least, we should automatically run tests on linux. Having access to other environments would be gravy.

### LINKED ISSUE #32: Flock supported in macports
Just a heads up more than an issue. I added flock to macports as of de9339a0adf26fa745aaeb1d1cf32985f3e3069f

Thus, if you want, you can add  
```
 port install flock
```
to the readme for macports.

### LINKED ISSUE #33: Compilation error with GCC
When building flock 0.4.0 with GCC 11.1.0, I get the following compilation error:

```
src/flock.c: In function ‘main’:
src/flock.c:323:32: error: format ‘%u’ expects argument of type ‘unsigned int’, but argument 2 has type ‘__suseconds_t’ {aka ‘long int’} [-Werror=format=]
  323 |                 printf("took %1u microseconds\n", (t_l_acq.tv_usec - t_l_req.tv_usec)); // not adding due to time constraints
      |                              ~~^                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                                |                                   |
      |                                unsigned int                        __suseconds_t {aka long int}
      |                              %1lu
cc1: all warnings being treated as errors
```

As the compiler recommends, it can be fixed by simply replacing `%1u` with `%1lu`. After making this change Flock was successfully built on my system.

### LINKED ISSUE #41: Unstated reliance on Bash Automated Testing System (bats)
`$ make check` fails as the code presumes the Bash Automated Testing System (bats) is installed.
Note the prerequisite somewhere or (even better) have `configure` denote its absence.

[test-suite.log](https://github.com/user-attachments/files/18265911/test-suite.log)

### PR DESCRIPTION
**Pull request by voltechs**
## Summary

- **GitHub Actions CI** with full test matrix: Ubuntu (GCC + Clang), macOS, and FreeBSD
- **Comprehensive test suite** expanded from ~30 timing-sensitive tests to 77 robust tests covering all flags, lock semantics, edge cases, and option combinations
- **Bug fixes** for GCC compilation errors, incorrect verbose timing, typos, and unvalidated user input
- **Build improvements** making ronn (man page generator) optional so `make && make install` works without extra dependencies

## Bug fixes

- `-h`/`--help` now exits 0 (was EX_USAGE/64) — fixes #15
- Fix `%1u` format specifier for `suseconds_t` causing GCC `-Werror` build failure — fixes #33 (supersedes #38)
- Fix verbose timing underflow: elapsed time was only subtracting `tv_usec`, producing wrong/negative values when lock acquisition crossed a second boundary
- Fix typo in error message: "waidpid" → "waitpid"
- Validate `-E` conflict exit code (was `atoi`, now rejects non-numeric and out-of-range values)
- Validate file descriptor argument in FD mode (previously `flock abc` silently locked stdin)

## Already fixed (can be closed)

- `-x` is already supported as a no-op (default is exclusive) — #9
- `-c` already wraps command with `$SHELL -c` as expected — #10

## Test suite

- Replaced exact timing comparisons (`[ "$result" = 0.05 ]`) with threshold-based helpers (`was_blocked`/`was_immediate`) — fixes #5, addresses #42
- Added `setup`/`teardown` with per-test lockfiles and background process cleanup
- New coverage: `-e` flag, `--nb` alias, `-c` shell features (pipes, variables, subst), child exit propagation, `-E` with `-n`/`-w`, `-o`/`--close`, `--verbose`, option combinations and flag precedence, file with spaces, nonexistent paths

## CI & build

- GitHub Actions replacing Travis — fixes #17
- Matrix: `ubuntu-latest` (gcc, clang), `macos-latest`, FreeBSD 15 via `vmactions/freebsd-vm`
- Man page build is now gracefully optional (warns instead of failing when `ronn` is absent) — fixes #41
- Actions pinned to v5 (Node.js 24)

## Documentation

- Added MacPorts install instructions to README — fixes #32 (supersedes #35)
- Documented `-c`, `-h`, `-V`, and `--verbose` options in man page

## Code quality

- Normalized indentation to tabs throughout `flock.c`
- Removed unnecessary `|=` for `fl.l_type` assignment in fcntl fallback
- Removed extraneous blank lines and fixed double-tab indentation

## Summary by CodeRabbit

* **New Features**
  * Added `-c/--command`, `-h/--help`, and `-V/--version` options; improved verbose timing output
  * Added CI that builds/tests across Linux, macOS, FreeBSD and generates man pages

* **Bug Fixes**
  * Fixed lock-type handling and corrected error messages
  * Tightened validation for exit-code and descriptor arguments

* **Documentation**
  * Expanded manpage OPTIONS and reordered verbosity docs
  * README notes that bats-core is required to run the test suite

* **Tests**
  * Significantly expanded and made tests timing-robust; improved setup/teardown and coverage

### CODE REVIEW

**Comment by voltechs** [comment]
Hi @josephholsten, I did a thing. Hope you find this helpful in modernizing `flock`. Thanks for writing it. Would love to see a version bump on `homebrew` in the near future.

**Comment by josephholsten** [comment]
hey there, it's been a while since I've worked on this so I've recruited some automatic code review.

I'm only concerned about the strtol() check and the  `|| true` swallowing a bats assertion.

File: README.md [hunk 1]
```diff
@@ -20,6 +20,10 @@ On macOS using Homebrew:
 
     brew install flock
 
+On macOS using MacPorts:
+
+    port install flock
```
[the remark this answers is not in the document]
Comment (voltechs) [reply README.md:25]: Personally, I would default to not using `sudo`, and let that be a discoverable necessity, if it comes to that. User's should always try to do things sans-sudo first, and sudo as a last resort. Anybody cognizant enough to be installing `flock` from MacPorts, let alone installing `flock` at all, can be trusted to know if/when they'll need `sudo`.

**Comment by voltechs** [comment]
Hey @josephholsten, thanks for the engagement. Sorry it took me so long to circle back to this. I worked through all of the code rabbits comments best I could, but it started to become whackamole. I'd fix everything on its list and the it'd find something else to complain about.

I think it did ultimately give it a pass. Let me know how you're feeling about this. I rebased onto master after your couple merges earlier, which addressed a few things the PR did as well.

**Comment by josephholsten** [comment]
Sorry about the delay in merging, you got the relevant stuff absolutely. It's trying to critique larger issues in the codebase, and those don't make sense in this changeset. You're awesome, thank you!

**Comment by voltechs** [comment]
Hey @josephholsten, thanks for merging #43! A few follow-ups now that it's in master.

**Issues #43 resolves that didn't auto-close:**

- **#42** (`make check` failures) — this was the old timing-sensitive suite failing on older macOS. The rewrite in #43 replaces the exact-timing assertions (`[ "$result" = 0.05 ]`) with poll-based lock checks and `was_blocked`/`was_immediate` thresholds, which is what #42 was hitting. I'd close it, ideally after @plroebuck confirms `make check` passes on current master.
- **#9** (`-x` as a no-op) — already satisfied: the default lock type is `LOCK_EX`, so `-x` is effectively a no-op, matching util-linux. Verified in `src/flock.c`.
- **#10** (`-c` differs from Linux) — already satisfied: `-c` runs the command via `$SHELL -c` (falling back to `/bin/sh`). Verified in `src/flock.c`.

I'd leave **#1** (smartos.bats — tests are more robust now but the omnios `flock(2)` gap is unfixed) and **#18** (flint/splint — I fixed the `-Werror=format` issue but haven't confirmed `make lint` runs clean) open.

**A release would be great.** The latest tag is v0.4.0 from 2020, and `configure.ac` still reads `0.4.0`, so none of #43 — the GCC build fix, CI, validation hardening, `-h`/`--help` to stdout, the rewritten tests — has shipped anywhere. `brew install flock` builds from the released tarball, so Homebrew users are still on 2020 code. Tagging a v0.5.0 (and bumping `configure.ac`) would get it all out, plus a homebrew-core formula bump.

Happy to help with any of it — release PR, changelog, or the homebrew-core bump. Just say the word.

**Comment by plroebuck** [comment]
So is there a `v0.5.0b` anywhere? Or you want me to download the codebase itself?

Or you could push your fork's codebase as a fake pull here...

Tell me what you would like me to do to test this, @voltechs

**Comment by voltechs** [comment]
@plroebuck there's no `v0.5.0b` let alone a `v0.5.0` because we're waiting on @josephholsten to cut a release. I've offered to help, but it's out of my hands at this point.
